compiler/AST/interfaces.cpp

```diff
@@ -0,0 +1,469 @@
+  // TODO update 'witnesses'
+  if (oldAst == implExpr)
+    implExpr = toImplementsExpr(newAst);
+  else if (oldAst == implBody)
+    implBody = toBlockStmt(newAst);
+  else
+    INT_FATAL(this, "Unexpected case in ImplementsStmt::replaceChild");
+}
+
+Expr* ImplementsStmt::getFirstExpr() {
+  return implExpr->getFirstExpr();
+}
+
+Expr* ImplementsStmt::getNextExpr(Expr* expr) {
+  if (expr == implExpr)
+    return implBody;
+  if (expr == implBody)
+    return this;
+
+  INT_FATAL(this, "Unexpected case in ImplementsStmt::getNextExpr");
+  return NULL;
+}
+
+
+/////////////////////////////////////////////////////////////////////////////
+// Support for normalize()
+
+
+//
+// Change proc cgFun(cgFormal: ?T)
+// to     proc cgFun(cgFormal: T)
+// plus add (DefExpr type T) to cgFun.constrainedTypes
+// where we expect 'T' to be bound by interface constraint(s).
+//
+void introduceConstrainedTypes(FnSymbol* fn) {
+  for_formals(formal, fn)
+   if (BlockStmt* typeExpr = formal->typeExpr)
+    if (DefExpr* def = toDefExpr(typeExpr->body.tail)) {
+      INT_ASSERT(formal->type == dtUnknown); //fyi
+      SET_LINENO(def);
+      Symbol* queryT = def->sym;
+      // introduce a ConstrainedType
+      TypeSymbol* CT = ConstrainedType::build(queryT->name);
+      fn->constrainedTypes.insertAtTail(new DefExpr(CT));
+
+      // replace queryT with CT throughout
+      formal->type = CT->type;
+      for_SymbolSymExprs(se, queryT)
+        se->setSymbol(CT);
+
+      // cleanup
+      def->remove();
+      if (! typeExpr->body.empty())
+        USR_FATAL(typeExpr, "this formal's type query expression"
+          " is currently not supported for constrained generic functions");
+      typeExpr->remove();
+    }
+}
+
+//
+// Change proc cgFun(cgFormal: IFC)
+// to     proc cgFun(cgFormal: t_IFC) where implements IFC(t_IFC)
+// plus add (DefExpr type t_IFC) to cgFun.constrainedTypes
+//
+Type* desugarInterfaceAsType(ArgSymbol* arg, SymExpr* se,
+                             InterfaceSymbol* isym) {
+  FnSymbol* parentFn = toFnSymbol(arg->defPoint->parentSymbol);
+  bool wasConstrainedGeneric = parentFn->isConstrainedGeneric();
+  // follows introduceConstrainedTypes()
+  SET_LINENO(se);
+
+  // introduce a ConstrainedType
+  TypeSymbol* CT = ConstrainedType::build(astr("t_", isym->name));
+  parentFn->constrainedTypes.insertAtTail(new DefExpr(CT));
+
+  // add an interface constraint
+  ImplementsExpr* ie = new ImplementsExpr(new SymExpr(isym));
+  ie->implActuals.insertAtTail(new SymExpr(CT));
+  parentFn->interfaceConstraints.insertAtTail(ie);
+
+  // catch up with moveGlobalDeclarationsToModuleScope()
+  if (!wasConstrainedGeneric)
+    if (ModuleSymbol* mod = toModuleSymbol(parentFn->defPoint->parentSymbol))
+      {
+        FnSymbol* initFn = mod->initFn;
+        // follows insertProxyRef()
+        // wass todo this will not retain the exact location,
+        // wass if needed, do desugarInterfaceAsType() earlier in normalize
```

Comment:
Seems like running it before moveGlobalDeclarationsToModuleScope would make this code simpler.